### ISSUE TITLE
Self documenting output files

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#2065]
It would be nice if our output files were self-documenting, with descriptions of what each variable means included in the netCDF files.

I've added the machinery for adding 'description's as attributes in #2064, which was fairly simple. [The meanings of the CVODE diagnostic variables was especially unclear to me, so I wanted to be able to copy a description from the CVODE docs to remind me what they were!] Especially for new users, this kind of documentation of all the variables BOUT++ outputs could be very helpful, I think. It's a big job to go through adding a new layer of documentation, but I don't see a section of the manual detailing the meaning of the variables in the output file - that could be a nice thing to add too!

The example in #2064 is maybe more complicated than most, because adding things to output in `Solver` is a bit convoluted... In other places it should just be a case of looking for things like
https://github.com/boutproject/BOUT-dev/blob/0448a8d12d5cf54b9dc75bcc7557af2e698dc86a/src/bout%2B%2B.cxx#L645
and adding a description
```
dump_file.add(simtime, "t_array", true, "time variable of the simulation");
```

**Comment by ZedThree** [comment]
Yes, completely onboard with this. I think ideally the output API would mirror the `Options` API for input. Ben made a start with the `OptionsNetCDF` class. I'd love for us to move to using that everywhere in 5.0.

**Comment by bendudson** [comment]
Awesome! Yes, this would be great. In addition to doc strings, I think we should try to standardise a way to attach units to output fields. Most simulations use normalised (dimensionless) values, so the metadata needs to be able to specify e.g.:

`Tnorm`   Units: eV
`Nnorm`   Units: m^-3
'P'       Normalised: e * Tnorm * Nnorm -> Pascals

This would enable post-processing tools to quickly translate our outputs into standard units.

**Comment by johnomotani** [comment]
Yes, units would be very nice to have too. Should be easy to add in a similar way; I guess 'description' and 'units' should be enough, at least for standardised things, and users can always use `setAttributes()` to add other custom attributes - so there aren't enough extra arguments to have to worry about the interface yet (?).

Support for units with `pint` in `xarray` is being worked on actively - hopefully it arrives in a version or two's time and then having unit attributes would be extra useful!

BTW I have some kind of memory bug in the code at the moment, valgrind complains and there are sometimes segfaults. I'll try to dig out, but for the moment this is another reason this is WIP.
